biscuit/test/Spec/Parser.hs

```diff
@@ -26,31 +30,31 @@ pk4 = fromJust $ parsePublicKeyHex "c6864578bc03596d52878bd70025ec966c95c60727cb
 pk5 = fromJust $ parsePublicKeyHex "a0d3dc7ab62a0a2732ba267e0d57894170458ec1659ca1226240b99764554a2e"
 
 parseTerm :: Text -> Either String Term
-parseTerm = parseOnly termParser
+parseTerm = (validationToEither . first show . substitutePTerm mempty) <=< run predicateTermParser
 
 parseTermQQ :: Text -> Either String QQTerm
-parseTermQQ = parseOnly termParser
+parseTermQQ = run predicateTermParser
 
 parsePredicate :: Text -> Either String Predicate
-parsePredicate = parseOnly predicateParser
+parsePredicate = (validationToEither . first show . substitutePredicate mempty) <=< run predicateParser
 
 parseRule :: Text -> Either String Rule
-parseRule = parseOnly ruleParser
+parseRule = (validationToEither . first show . substituteRule mempty mempty) <=< run ruleParser
 
 parseExpression :: Text -> Either String Expression
-parseExpression = parseOnly expressionParser
+parseExpression = (validationToEither . first show . substituteExpression mempty) <=< run expressionParser
 
 parseCheck :: Text -> Either String Check
-parseCheck = parseOnly checkParser
+parseCheck = (validationToEither . first show . substituteCheck mempty mempty) <=< run checkParser
 
 parseAuthorizer :: Text -> Either String Authorizer
-parseAuthorizer = parseOnly authorizerParser
+parseAuthorizer = (validationToEither . first show . substituteAuthorizer mempty mempty) <=< run authorizerParser
 
 parsePolicy :: Text -> Either String Policy
-parsePolicy = parseOnly policyParser
+parsePolicy = (validationToEither . first show . substitutePolicy mempty mempty) <=< run policyParser
 
 parseBlock :: Text -> Either String Block
-parseBlock = parseOnly blockParser
+parseBlock = (validationToEither . first show . substituteBlock mempty mempty) <=< run blockParser
```

Comment:
This diff is screaming outloud: don't repeat yourself please 😬

Reply:
Ha, this part was written in anger. I'll factor the helper out